### ISSUE TITLE
Jira Export: System.NullReferenceException Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by hochstetter** [solidify/jira-azuredevops-migrator#328]
Hello,

i need a little bit of help using your great tool and I didn't find the exact problem in your issue tracker. I translated the german part of the error message `Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.` with the title message. I think it is correct because I saw that english message in other issues here.

Kind regards,
Thomas Hochstetter

**Describe the bug**
I want to retrieve the Jira Items of a on-premise Jira server and on some storys I get an error. After processing 12 items successful, the next one get the following error. 

If I exclude the error causing stories in the Jira query in the config file (like _key not in WXYZ-1234_) the other items are exported fine. The error seems to be in the comment section of that story if I read the error log correct? Is that right? But I can open the Jira story successfully in the browser without an error.

**To Reproduce**
It's a private Jira server, so I don't have steps to reproduce. I also tried the force parameter and the continue parameter but that didn't help.

1. I start the app with a command line like this `jira-export.exe -u xyz@xyz.de -p 4TRUTZGFHFm7EEDF --url https://xyz.com/jira --config config-scrum-xxxx.json`
2. The following log is created.

```
====================================================================
Jira Export Log
====================================================================
Tool version : 2.3.117
Start time   : 22.02.2021 16:22:12
Telemetry    : Enabled
Session id   : xxxx
Tool user    : **********
Config       : config-scrum-xxxx.json
Force        : yes
Log level    : Info
Machine      : 1234567
System       : Microsoft Windows 10.0.18363
Jira url     : https://xyz.com/jira
Jira user    : x
Jira version : 8.5.7
Jira type    : Server
====================================================================
[I][16:22:10] Connecting to Jira...
[I][16:22:10] Retrieving Jira fields...
[I][16:22:11] Retrieving Jira link types...
[I][16:22:12] Export started. Exporting 46 items.
[I][16:22:13] Initializing Jira field mapping...
[I][16:22:14] Processing 1/46 - 'WXYZ-4457'.

....

[I][16:22:40] Processing 11/46 - 'WXYZ-3662'.
[W][16:22:40] Could not find user '22811' identity in user map. Using default identity 'Test'.
[I][16:22:42] Processing 12/46 - 'WXYZ-2719'.
[E][16:22:42] Unexpected migration error.
[System.NullReferenceException] System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei JiraExport.JiraItem.<>c__DisplayClass2_0.<BuildCommentRevisions>b__0(Comment c, Int32 i)
   bei System.Linq.Enumerable.<SelectIterator>d__5`2.MoveNext()
   bei System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   bei System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   bei JiraExport.JiraItem.BuildCommentRevisions(JiraItem jiraItem, JiraProvider jiraProvider)
   bei JiraExport.JiraItem.BuildRevisions(JiraItem jiraItem, JiraProvider jiraProvider)
   bei JiraExport.JiraItem.CreateFromRest(String issueKey, JiraProvider jiraProvider)
   bei JiraExport.JiraProvider.<EnumerateIssues>d__23.MoveNext()
   bei JiraExport.JiraCommandLine.ExecuteMigration(CommandOption user, CommandOption password, CommandOption url, CommandOption configFile, Boolean forceFresh, CommandOption continueOnCritical): 
   bei JiraExport.JiraItem.<>c__DisplayClass2_0.<BuildCommentRevisions>b__0(Comment c, Int32 i)
   bei System.Linq.Enumerable.<SelectIterator>d__5`2.MoveNext()
   bei System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   bei System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   bei JiraExport.JiraItem.BuildCommentRevisions(JiraItem jiraItem, JiraProvider jiraProvider)
   bei JiraExport.JiraItem.BuildRevisions(JiraItem jiraItem, JiraProvider jiraProvider)
   bei JiraExport.JiraItem.CreateFromRest(String issueKey, JiraProvider jiraProvider)
   bei JiraExport.JiraProvider.<EnumerateIssues>d__23.MoveNext()
   bei JiraExport.JiraCommandLine.ExecuteMigration(CommandOption user, CommandOption password, CommandOption url, CommandOption configFile, Boolean forceFresh, CommandOption continueOnCritical)
[I][16:22:42] Export complete. Exported 46 items (1 errors, 36 warnings) in 00:00:32.
```

**Expected behavior**
A better error message would be nice. I don't know what is wrong with the story. If I would know what is wrong, maybe I could edit the story to get it passed without error. Or a command line flag that will export everything that is possible of faulty Jira Elements that produce an error. Maybe I misused the continue command line parameter. What is the correct syntax? It is not documented in https://github.com/solidify/jira-azuredevops-migrator/blob/master/docs/jira-export.md .

**Desktop (please complete the following information):**
 - OS: Windows 10.0.18363 (German)
 - Browser Chrome 88.0.4324.182

**Config File**
I don't know if it helps.

```
{
  "source-project": "WXYZ",
  "target-project": "WXYZ",
  "query": "project = WXYZ and key not in (WXYZ-1346,WXYZ-5635) ORDER BY Rank ASC",
  "using-jira-cloud": true,
  "workspace": "C:\\JiraAzure\\JiraExport\\",
  "epic-link-field": "Epic Link",
  "sprint-field": "Sprint",
  "download-options": 7,
  "batch-size": 20,
  "log-level": "Info",
  "attachment-folder": "Attachments",
  "user-mapping-file": "C:\\JiraAzure\\jira-azuredevops-migrator-2.3.117\\0_user_mapping.txt",
  "base-area-path": "Migrated",
  "base-iteration-path": "Migrated",
  "ignore-failed-links": true,
  "process-template": "Scrum",
  "link-map": {
    "link": [
      {
        "source": "Epic",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Parent",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Relates",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "Duplicate",
        "target": "System.LinkTypes.Duplicate-Forward"
      }
    ]
  },
  "type-map": {
    "type": [
      {
        "source": "Feature",
        "target": "Feature"
      },
      {
        "source": "Epic",
        "target": "Epic"
      },
      {
        "source": "Story",
        "target": "Product Backlog Item"
      },
      {
        "source": "Bug",
        "target": "Bug"
      },
      {
        "source": "Task",
        "target": "Product Backlog Item"
      },
      {
        "source": "Sub-Task",
        "target": "Task"
      },
      {
        "source": "Sub-task",
        "target": "Task"
      }
    ]
  },
  "field-map": {
    "field": [
      {
        "source": "summary",
        "target": "System.Title",
        "mapper": "MapTitle"
      },
      {
        "source": "assignee",
        "target": "System.AssignedTo",
        "mapper": "MapUser"
      },
      {
        "source": "description",
        "target": "System.Description",
        "mapper": "MapRendered"
      },
      {
        "source": "priority",
        "target": "Microsoft.VSTS.Common.Priority",
        "mapping": {
          "values": [
            {
              "source": "Blocker",
              "target": "1"
            },
            {
              "source": "Critical",
              "target": "1"
            },
            {
              "source": "Kritisch",
              "target": "1"
            },			
            {
              "source": "Highest",
              "target": "1"
            },
            {
              "source": "Major",
              "target": "2"
            },
            {
              "source": "High",
              "target": "2"
            },
            {
              "source": "Schwer",
              "target": "2"
            },			
            {
              "source": "Medium",
              "target": "3"
            },
            {
              "source": "Low",
              "target": "3"
            },
            {
              "source": "Lowest",
              "target": "4"
            },
            {
              "source": "Minor",
              "target": "4"
            },
            {
              "source": "Trivial",
              "target": "4"
            },
            {
              "source": "Unwesentlich",
              "target": "4"
            },
            {
              "source": "None",
              "target": ""
            }			
          ]
        }
      },
      {
        "source": "Sprint",
        "source-type": "name",
        "target": "System.IterationPath",
        "mapper": "MapSprint"
      },
      {
        "source": "labels",
        "target": "System.Tags",
        "mapper": "MapTags"
      },
      {
        "source": "comment",
        "target": "System.History",
        "mapper": "MapRendered"
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Task",
        "mapping": {
          "values": [
            {
              "source": "To Do",
              "target": "New"
            },
            {
              "source": "Done",
              "target": "Done"
            },
            {
              "source": "In Progress",
              "target": "Committed"
            },
			            {
              "source": "New",
              "target": "New"
            },
            {
              "source": "Fertig",
              "target": "Approved"
            },
            {
              "source": "In Arbeit",
              "target": "Committed"
            },
            {
              "source": "Geschlossen",
              "target": "Done"
            },
            {
              "source": "On Hold",
              "target": "New"
            },
            {
              "source": "Ready",
              "target": "Approved"
            },
            {
              "source": "Obsolete",
              "target": "Done"
            },
            {
              "source": "Review",
              "target": "Committed"
            },
            {
              "source": "Offen",
              "target": "Approved"
            },
            {
              "source": "Analysis",
              "target": "New"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Bug,Product Backlog Item",
        "mapping": {
          "values": [
            {
              "source": "To Do",
              "target": "New"
            },
            {
              "source": "Done",
              "target": "Done"
            },
            {
              "source": "In Progress",
              "target": "Committed"
            },
			            {
              "source": "New",
              "target": "New"
            },
            {
              "source": "Fertig",
              "target": "Approved"
            },
            {
              "source": "In Arbeit",
              "target": "Committed"
            },
            {
              "source": "Geschlossen",
              "target": "Done"
            },
            {
              "source": "On Hold",
              "target": "New"
            },
            {
              "source": "Ready",
              "target": "Approved"
            },
            {
              "source": "Obsolete",
              "target": "Done"
            },
            {
              "source": "Review",
              "target": "Committed"
            },
            {
              "source": "Offen",
              "target": "Approved"
            },
            {
              "source": "Analysis",
              "target": "New"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Epic,Feature",
        "mapping": {
          "values": [
            {
              "source": "To Do",
              "target": "New"
            },			
            {
              "source": "Done",
              "target": "Done"
            },
            {
              "source": "In Progress",
              "target": "Committed"
            },
			{
              "source": "New",
              "target": "New"
            },
            {
              "source": "Fertig",
              "target": "Approved"
            },
            {
              "source": "In Arbeit",
              "target": "Committed"
            },
            {
              "source": "Geschlossen",
              "target": "Done"
            },
            {
              "source": "On Hold",
              "target": "New"
            },
            {
              "source": "Ready",
              "target": "Approved"
            },
            {
              "source": "Obsolete",
              "target": "Done"
            },
            {
              "source": "Review",
              "target": "Committed"
            },
            {
              "source": "Offen",
              "target": "Approved"
            },
            {
              "source": "Analysis",
              "target": "New"
            }
          ]
        }
      },
      {
        "source": "Story Points",
        "source-type": "name",
        "target": "Microsoft.VSTS.Scheduling.Effort",
        "not-for": "Task"
      },
      {
        "source": "timeestimate",
        "target": "Microsoft.VSTS.Scheduling.RemainingWork",
        "mapper": "MapRemainingWork",
        "for": "Bug,Task"
      },
      {
        "source": "description",
        "target": "Microsoft.VSTS.TCM.ReproSteps",
        "for": "Bug"
      }
    ]
  }
}
```

AB#5950

**Comment by hochstetter** [comment]
I identified the problem using your open source code 👍. There is no not null check in `JiraItem.BuildCommentRevisions()` for `c.AuthorUser`. In the affected Jira Server there are some items with c.AuthorUser = null

```C#
                return new JiraRevision(jiraItem)
                {
                    Author = c.AuthorUser.Username ?? GetAuthorIdentityOrDefault(c.AuthorUser.AccountId),
                    Time = c.CreatedDate.Value,
                    Fields = new Dictionary<string, object>() { { "comment", c.Body }, { "comment$Rendered", rc.Value<string>() } },
                    AttachmentActions = new List<RevisionAction<JiraAttachment>>(),
                    LinkActions = new List<RevisionAction<JiraLink>>()
                };
```

I changed the code to something like this.

```C#
        private static List<JiraRevision> BuildCommentRevisions(JiraItem jiraItem, JiraProvider jiraProvider)
        {
            var renderedFields = jiraItem.RemoteIssue.SelectToken("$.renderedFields.comment.comments");
            var comments = jiraProvider.Jira.Issues.GetCommentsAsync(jiraItem.Key).Result;
            return comments.Select((c, i) =>
            {
                var rc = renderedFields.SelectToken($"$.[{i}].body");

                var theAuthor = "noAuthorDefined";
                if (c.AuthorUser is null)
                {
                    Logger.Log(LogLevel.Warning, $"c.AuthorUser is null in comment revision for jiraItem.Key: '{ jiraItem.Key}'. Using noAuthorDefined as author. ");
                }
                else
                {
                    if (c.AuthorUser.Username is null)
                    {
                        theAuthor = GetAuthorIdentityOrDefault(c.AuthorUser.AccountId);
                    }
                    else
                    {
                        theAuthor = c.AuthorUser.Username;
                    }
                }
                Logger.Log(LogLevel.Warning, $"Der Autor wird jetzt geschrieben: '{theAuthor}'");

                return new JiraRevision(jiraItem)
                {
                    Author = theAuthor,
                    Time = c.CreatedDate.Value,
                    Fields = new Dictionary<string, object>() { { "comment", c.Body }, { "comment$Rendered", rc.Value<string>() } },
                    AttachmentActions = new List<RevisionAction<JiraAttachment>>(),
                    LinkActions = new List<RevisionAction<JiraLink>>()
                };
            }).ToList();
        }
```

I tried to push these changes to your git repo but I couldn't. Visual Studio gave this error:
```
Pushing fix/BuildCommentRevisions_No_Authoruser
Error encountered while pushing branch to the remote repository: Git failed with a fatal error.
unable to access 'https://github.com/solidify/jira-azuredevops-migrator/': The requested URL returned error: 403
Pushing to https://github.com/solidify/jira-azuredevops-migrator

Failed to push the branch to the remote repository. See the Output window for more details.
```
This is the first time I used Github, Visual Studio Community 2019 and C#. That's why I don't know how to do that. For the build process I also needed some help by my colleague.

**I think** the null error can also happen to  `c.CreatedDate` and `rc`?

**Comment by tung-nt-niteco** [comment]
Cool, thanks.

**Comment by Alexander-Hjelm** [comment]
Will merge this change at a later stage. Thank you!

**Comment by Alexander-Hjelm** [comment]
Fixed by #523
